modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -0,0 +1,1788 @@
+        ServiceProcessorCommonDiscoveryData clusterData = (ServiceProcessorCommonDiscoveryData)data.commonData();
+
+        for (ServiceInfo desc : clusterData.registeredServices())
+            registeredServices.put(desc.serviceId(), desc);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void collectJoiningNodeData(DiscoveryDataBag dataBag) {
+        ArrayList<ServiceInfo> staticServicesInfo = staticallyConfiguredServices(true);
+
+        dataBag.addJoiningNodeData(SERVICE_PROC.ordinal(), new ServiceProcessorJoinNodeDiscoveryData(staticServicesInfo));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onJoiningNodeDataReceived(DiscoveryDataBag.JoiningNodeDiscoveryData data) {
+        if (data.joiningNodeData() == null)
+            return;
+
+        ServiceProcessorJoinNodeDiscoveryData joinData = (ServiceProcessorJoinNodeDiscoveryData)data.joiningNodeData();
+
+        for (ServiceInfo desc : joinData.services()) {
+            assert desc.topologySnapshot().isEmpty();
+
+            ServiceInfo oldDesc = registeredServices.get(desc.serviceId());
+
+            if (oldDesc != null) { // In case of a collision of IgniteUuid.randomUuid() (almost impossible case)
+                U.warn(log, "Failed to register service configuration received from joining node : " +
+                    "[nodeId=" + data.joiningNodeId() + ", cfgName=" + desc.name() + "]. " +
+                    "Service with the same service id already exists, cfg=" + oldDesc.configuration());
+
+                continue;
+            }
+
+            oldDesc = lookupInRegisteredServices(desc.name());
+
+            if (oldDesc == null) {
+                registeredServices.put(desc.serviceId(), desc);
+
+                continue;
+            }
+
+            if (oldDesc.configuration().equalsIgnoreNodeFilter(desc.configuration())) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Ignore service configuration received from joining node : " +
+                        "[nodeId=" + data.joiningNodeId() + ", cfgName=" + desc.name() + "]. " +
+                        "The same service configuration already registered.");
+                }
+            }
+            else {
+                U.warn(log, "Failed to register service configuration received from joining node : " +
+                    "[nodeId=" + data.joiningNodeId() + ", cfgName=" + desc.name() + "]. " +
+                    "Service already exists with different configuration, cfg=" + desc.configuration());
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Nullable @Override public DiscoveryDataExchangeType discoveryDataType() {
+        return SERVICE_PROC;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onActivate(GridKernalContext kctx) {
+        // No-op.
+    }
+
+    /**
+     * Invokes from services deployment worker.
+     * <p/>
+     * {@inheritDoc}
+     */
+    @Override public void onDeActivate(GridKernalContext kctx) {
```

Comment:
Is this actually called? I see `onDeActivate` being called exclusively for `GridServiceProcessor` in some other places.

Reply:
Yes, it's called from own deployment manager.
`GridServiceProcessor` is called in PME flow, because it is based on an internal cache, that means it has to be deactivated with caches.

Reply:
When will this method be called if not during PME? I'd expect all services to be cancelled by the time `cluster().active(false)` returns. Is it the case?

Reply:
Sure, it is. `ServiceDeploymentManager` listens to `ChangeGlobalStateMessage` to perform activation/deactivation.